enterprise/kernel/src/main/java/org/neo4j/kernel/enterprise/builtinprocs/BuiltInProcedures.java

```diff
@@ -54,56 +70,258 @@
+            .map( this::queryStatusResult );
+    }
+
+    @Description( "Kill all transactions executing the query with the given query id." )
+    @Procedure( name = "dbms.killQuery", mode = DBMS )
+    public Stream<QueryTerminationResult> killQuery( @Name( "id" ) String idText )
+            throws InvalidArgumentsException, IOException
+    {
+        long queryId = parseQueryId( idText ).kernelQueryId();
+
+        Set<Pair<KernelTransactionHandle,ExecutingQuery>> executingQueries =
+            getKernelTransactions().activeTransactions( tx -> executingQueriesWithId( queryId, tx ) );
+
+        return executingQueries
+            .stream()
+            .map(this::killQueryTransaction);
+    }
+
+    @Description( "Kill all transactions executing a query with any of the given query ids." )
+    @Procedure( name = "dbms.killQueries", mode = DBMS )
+    public Stream<QueryTerminationResult> killQueries( @Name( "ids" ) List<String> idTexts )
+            throws InvalidArgumentsException, IOException
+    {
+        Set<Long> queryIds =
+            idTexts.stream()
+                   .map( QueryId::parseQueryId )
+                   .map( QueryId::kernelQueryId )
+                   .collect( Collectors.toSet() );
+
+        Set<Pair<KernelTransactionHandle,ExecutingQuery>> executingQueries =
+                getKernelTransactions().activeTransactions( tx -> executingQueriesWithIds( queryIds, tx ) );
+
+        return executingQueries
+                .stream()
+                .map(this::killQueryTransaction);
+    }
+
+    private Stream<ExecutingQuery> executingQueriesWithIds( Set<Long> ids, KernelTransactionHandle txHandle )
+    {
+        return txHandle.executingQueries().filter( q -> ids.contains( q.kernelQueryId() ) );
+    }
+
+    private Stream<ExecutingQuery> executingQueriesWithId( long id, KernelTransactionHandle txHandle )
+    {
+        return txHandle.executingQueries().filter( q -> q.kernelQueryId() == id );
+    }
+
+    private QueryTerminationResult killQueryTransaction( Pair<KernelTransactionHandle, ExecutingQuery> pair )
+    {
+        ExecutingQuery query = pair.other();
+        if ( isAdminEnterpriseAuthSubject() || authSubject.hasUsername( query.username() ) )
+        {
+            pair.first().markForTermination( Status.Transaction.Terminated );
+            return new QueryTerminationResult( queryId( query.kernelQueryId() ), query.username() );
+        }
+        else
+        {
+            throw new AuthorizationViolationException( PERMISSION_DENIED );
+        }
     }
 
     private KernelTransactions getKernelTransactions()
     {
         return resolver.resolveDependency( KernelTransactions.class );
     }
 
-    private boolean isAdmin()
+    // ----------------- helpers ---------------------
+
+    private Stream<TransactionTerminationResult> terminateTransactionsForValidUser( String username )
+    {
+        long terminatedCount = 0;
```

Comment:
hm looks like a place where stream api should be used to count terminatedCount instead of for loop